src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -310,29 +316,53 @@ public CachingRebufferer(ChunkReader file)
+        public BufferHolder rebuffer(long position)
         {
-            int spin = 0;
             try
             {
                 long pageAlignedPos = position & alignmentMask;
-                Buffer buf;
-                Key key = new Key(source, internedPath, pageAlignedPos);
+                BufferHolder buf;
+                Key chunkKey = new Key(source, internedPath, pageAlignedPos);
+
+                int spin = 0;
+                //There is a small window when a released buffer/invalidated chunk
+                //is still in the cache. In this case it will return null
+                //so we spin loop while waiting for the cache to re-populate
                 while (true)
                 {
-                    buf = cache.get(key).reference();
+                    Buffer chunk;
+                    // Using cache.get(k, compute) results in lots of allocation, rather risk the unlikely race...
+                    CompletableFuture<Buffer> cachedValue = cache.getIfPresent(chunkKey);
+                    if (cachedValue == null)
+                    {
+                        // this blocking read might compete with an another read, but the race is benign (under the assumption that caching is valid)
+                        CompletableFuture<Buffer> entry = new CompletableFuture<>();
+                        // Put the future in the cache first so the window of racing (and loading the same chunk twice) is small
+                        cache.put(chunkKey, entry);
+
+                        try
+                        {
+                            chunk = load(chunkKey);
+                        }
+                        catch (Throwable t)
+                        {
+                            // also signal other waiting readers
+                            entry.completeExceptionally(t);
+                            throw t;
+                        }
+                        entry.complete(chunk);
+                    }
+                    else
+                    {
+                        chunk = cachedValue.join();
+                    }
+
+                    buf = chunk.reference();
                     if (buf != null)
                         return buf;
 
-                    if (++spin == 1000)
-                    {
-                        String msg = String.format("Could not acquire a reference to for %s after 1000 attempts. " +
-                                                   "This is likely due to the chunk cache being too small for the " +
-                                                   "number of concurrently running requests.", key);
-                        throw new RuntimeException(msg);
-                        // Note: this might also be caused by reference counting errors, especially double release of
-                        // chunks.
-                    }
+                    if (++spin == 1024)
+                        logger.error("Spinning for {}", chunkKey);
```

Comment:
Why do we need to remove `throw new RuntimeException(msg);`? It seems like that is important for preventing infinite loops.

Reply:
I don't like this infinite loop either here. If there is a bug elsewhere, e.g. double release, it would be spinning forever.

Reply:
I have added back the "throw"